### PR TITLE
[MLIR][LLVM] Add `llvm.experimental.constrained.fptrunc` operation

### PR DESCRIPTION
**Pull request by victor-eds**
Add operation mapping to the LLVM
`llvm.experimental.constrained.fptrunc.*` intrinsic.

The new operation implements the new
`LLVM::ExceptionBehaviorOpInterface` and
`LLVM::RoundingModeOpInterface` interfaces.

### CODE REVIEW

File: mlir/lib/Target/LLVMIR/ModuleImport.cpp
Comment (gysit) [inline mlir/lib/Target/LLVMIR/ModuleImport.cpp:1300]: The preferred way for importing attributes are the matcher functions we introduced for other attributes. We then added a tablegen shortcut so that the attributes can be parsed directly in tablegen.

A good example is `matchLabelAttr`: 
```
DILabelAttr ModuleImport::matchLabelAttr(llvm::Value *value) {
  auto *nodeAsVal = cast<llvm::MetadataAsValue>(value);
  auto *node = cast<llvm::DILabel>(nodeAsVal->getMetadata());
  return debugImporter->translate(node);
}
```
It is then add to LLVMIRConversionGen.cpp:
```
    } else if (name == "_label_attr") {
      bs << "moduleImport.matchLabelAttr";
```

Which then can be used inside the mlir builder:
```
  let mlirBuilder = [{
    DILabelAttr labelAttr = $_label_attr($label);
    // Drop the intrinsic if the label translation fails due to cylic metadata.
    if (!labelAttr)
      return success();
    $_op = $_builder.create<$_qualCppClassName>($_location, labelAttr);
  }];
```
That way the main import logic can hopefully be moved into tablegen?

Would something similar to matchLabelAttr work for you as well?

**Review by gysit — COMMENTED** [review]
Thanks for working on this!

I added some suggestions that are a bit closer to the solutions we took in other places. Specifically when importing intrinsic with metadata arguments. If the mechanism also works in your case, it would be great to follow the same idea.

File: mlir/include/mlir/Dialect/LLVMIR/LLVMEnums.td
Comment (zero9178) [inline mlir/include/mlir/Dialect/LLVMIR/LLVMEnums.td:745]: At the risk of bikeshedding, could these mention `FP` somewhere in the name? (e.g. `FPExceptionBehaviour`). At first glance I was wondering what a floating point intrinsic has to do with exceptions (as in itanium/C++ exceptions) :sweat_smile:

File: mlir/include/mlir/Dialect/LLVMIR/LLVMInterfaces.td
Comment (victor-eds) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMInterfaces.td]: I tried this, but the setter causes a compilation issue. There is a collision with a concrete operation member function with a different input type and the call is causing a bad signature error due to how MLIR trait classes are implemented. This is working for other cases as the attribute class has an implicit conversion operator to the underlying value. As the attributes we are dealing with here are enum attributes, we cannot add that member to the class via `extraClassDeclarations`. 

I'd keep this as is as I'd rather benefit from `LLVM_Enum` tablegen goodies. Are you fine with keeping this as is? I see other attributes with potentially the same issue that have to take the same compromise as here.

**Comment by victor-eds** [comment]
@zero9178 @gysit thanks for your comments! I hope I've addressed them all

**Review by zero9178 — APPROVED** [review]
LGTM, so if @gysit is happy then so am I :slightly_smiling_face:

File: mlir/include/mlir/Dialect/LLVMIR/LLVMIntrinsicOps.td
Comment (gysit) [inline mlir/include/mlir/Dialect/LLVMIR/LLVMIntrinsicOps.td:375]: nit: I would move the two asserts (for the exception and rounding attribute) into the ModuleImport.cpp similar to `matchFloatAttr` etc.

File: mlir/include/mlir/Dialect/LLVMIR/LLVMIntrinsicOps.td
Comment (gysit) [inline mlir/include/mlir/Dialect/LLVMIR/LLVMIntrinsicOps.td:341]:
```suggestion
    // Take into account overloaded result type.
```
Ultra nit: the comments here and below should end with a dot.

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (gysit) [inline mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:1724]:
```suggestion
```

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (gysit) [inline mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:1730]:
```suggestion
```

File: mlir/lib/Target/LLVMIR/ModuleImport.cpp
Comment (gysit) [inline mlir/lib/Target/LLVMIR/ModuleImport.cpp:1313]: Probably the assert would check if the optional has a value here? Or can the later call to `convertFPExceptionBehaviorFromLLVM` also go wrong?

**Review by gysit — APPROVED** [review]
Thanks for the changes!

LGTM modulo some minor comments.

Its a shame these intrinsics seem to exists with all kinds of argument number and rounding mode configurations. Otherwise, it may have been nice to factor out base classes unary / binary etc to make things a bit less complex...

File: mlir/include/mlir/Dialect/LLVMIR/LLVMInterfaces.td
Comment (gysit) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMInterfaces.td]: Uff ok I was not aware of these details. Then I guess it is fine to fallback to the string based solutions. As you said this is already used in other places...

**Comment by victor-eds** [comment]
> Its a shame these intrinsics seem to exists with all kinds of argument number and rounding mode configurations. Otherwise, it may have been nice to factor out base classes unary / binary etc to make things a bit less complex...

I guess we could have some kind of argument for the base intrinsic class in a similar way to immediate values, but that'd require further modifications.

**Comment by gysit** [comment]
> I guess we could have some kind of argument for the base intrinsic class in a similar way to immediate values, but that'd require further modifications. I don't think it'd be 100 % automatic anyway...

A solution where the arguments are defined in a derived class being the actual intrinsic or some unary / binary etc base class could be nice. Especially since that may allow us to properly type the intrinsic operands etc. However, with tablegen it is really hard to guess how complex things are in the end. If you have a good idea you are very welcome to give it a shot.

Otherwise it is also ok to land and follow up if needed (e.g. after adding more intrinsics).

**Comment by victor-eds** [comment]
> > I guess we could have some kind of argument for the base intrinsic class in a similar way to immediate values, but that'd require further modifications. I don't think it'd be 100 % automatic anyway...
> 
> A solution where the arguments are defined in a derived class being the actual intrinsic or some unary / binary etc base class could be nice. Especially since that may allow us to properly type the intrinsic operands etc. However, with tablegen it is really hard to guess how complex things are in the end. If you have a good idea you are very welcome to give it a shot.
> 
> Otherwise it is also ok to land and follow up if needed (e.g. after adding more intrinsics).

Yes, that's more or less what I was thinking of. I'm depending on this PR for further work in other dialects, so I'd rather get this merged and give this a shot in the future. I'll wait till CI passes. Thanks a lot!

**Pull request merged**
